scipy/fft/_pocketfft/scipy_features.h

```diff
@@ -0,0 +1,233 @@
+                                const pocketfft::shape_t & axes)
+  {
+  if (norm == norm_t::none) return T(1);
+
+  T N(1);
+  for (auto a : axes)
+    N *= shape[a];
+
+  if (norm == norm_t::size) return T(1) / N;
+  if (norm == norm_t::ortho) return T(1) / sqrt(N);
+
+  throw std::invalid_argument("invalid normalization type");
+  }
+
+
+pybind11::array asfarray(const pybind11::array & a, bool & inplace)
+  {
+    auto dtype = a.dtype();
+    switch (dtype.kind())
+      {
+      case 'c': // complex
+        return a;
+
+      case 'f': // floating
+        {
+        static const auto f16 = pybind11::dtype("float16");
+        if (!dtype.is(f16))
+          return a;
+
+        inplace = true;
+        return a.cast<pybind11::array_t<float>>();
+        }
+
+      default:
+        {
+        inplace = true;
+        return a.cast<pybind11::array_t<double>>();
+        }
+      }
+  }
+
+pybind11::array asfarray(const pybind11::array & a)
+  {
+  bool ignored;
+  return asfarray(a, ignored);
+  }
+
+
+namespace scipy {
+// Copied from pypocketfft.cxx
+pocketfft::shape_t copy_shape(const pybind11::array &arr)
+{
+  pocketfft::shape_t res(size_t(arr.ndim()));
+  for (size_t i=0; i<res.size(); ++i)
+    res[i] = size_t(arr.shape(int(i)));
+  return res;
+}
+
+// Copied from pypocketfft.cxx
+pocketfft::stride_t copy_strides(const pybind11::array &arr)
+{
+  pocketfft::stride_t res(size_t(arr.ndim()));
+  for (size_t i=0; i<res.size(); ++i)
+    res[i] = arr.strides(int(i));
+  return res;
+}
+}
+
+template<typename T> pybind11::array xfftn_real(const pybind11::array &in,
+  const pocketfft::shape_t & axes, norm_t norm, bool /*inplace*/, bool fwd,
+  size_t nthreads)
+  {
+  auto dims_in = scipy::copy_shape(in);
+  auto dims_out = dims_in;
+
+  const auto full_len = dims_in[axes.back()];
+  const auto rfft_len = full_len / 2 + 1;
+
+  dims_out[axes.back()] = rfft_len;
+  pybind11::array_t<std::complex<T>> res(dims_in);
+
+  // Perform real fft on the last axis
+  auto strides_in = scipy::copy_strides(in);
+  auto strides_out = scipy::copy_strides(res);
+  auto * in_data = reinterpret_cast<const T *>(in.data());
+  auto * data = res.mutable_data();
+  const auto size = static_cast<size_t>(res.size());
+
+  {
+  pybind11::gil_scoped_release release;
+  auto fct = norm_fct<T>(norm, dims_in, axes);
+
+  pocketfft::r2c(dims_in, strides_in, strides_out, axes.back(), in_data, data,
+                 fct, nthreads);
+
+  // For real input, ifft is the conjugate of the fft
+  if (!fwd)
+    {
+    for (size_t i = 0; i < size; i++)
+      data[i] = conj(data[i]);
+    }
+
+  // Perform complex (i)fft on remaining axes
+  if (axes.size() > 1)
+    {
+    pocketfft::shape_t new_axes(axes.begin(), axes.end()-1);
+    pocketfft::c2c(dims_out, strides_out, strides_out, new_axes, fwd, data, data,
+                   T(1), nthreads);
+    }
+
+  // Convert strides from bytes to objects
+  for (auto & s : strides_out)
+    s /= sizeof(std::complex<T>);
```

Comment:
I'm not sure that this will work in all situations. It's not unthinkable that the stride is not cleany divisible by sizeof(std::complex<T>) (e.g. in structured arrays).

Reply:
Nevermind ... I realized that you are computing the strides for an array that you just created yourself; that should always be safe. Problems will start once `inplace` or `out` arguments are used, since you no longer have control over the strides then.